### PR TITLE
feature/generic-core-update

### PR DESCRIPTION
**Pull request by janbiasi**
## Important changes summarized:

- **ENHANCEMENT**: Core DOM utils does now support generic parameters
    - `getComponentByDomNode`, `getComponentByDomNodeAsync` and `findComponents`
- **ADDITION**: Anchor attributes (`_gondel_g` or `_gondelA_g`) are now symbols (if supported) and can be generated with a new function called `getGondelAttribute(ns: string = 'g', addition?: string)`
   - Some functions are now even smaller because component class references are passed instead of the name so the reference must not be resolved manually via registry.
   - Anchor attributes inspired by [RXJS (rxSubscriber)](https://github.com/ReactiveX/rxjs/blob/master/src/internal/symbol/rxSubscriber.ts) implementation.
- **BREAKING**: DOM attribute mappings are now defined inside the component class and not the decorator, so each component can be identified at any time to which attribute the class belongs (`@Component('Button')` does not work anymore). The new way to define the mapping value is to add the static class property `public static componentName = 'Button'`.

### Other changes / notes

- Updated React example (`react-lazy`) to show how to work with child components inside react.
- Angular example can't be added atm. because its lack of support for monorepos (short insights for owners here: https://gist.github.com/janbiasi/8fb106bd20ea8e366aecbadb0e9c0126) 

## What are the benefits from this change?
More secure way to attach the component to DOM nodes (Symbols) inspired by bigger frameworks such as RXJS, Angular amm. The component annotation is now only used for the registration of the component, the mapping definition of the component is now located inside the class as static value. So in each core method where you had to pass a string with the name to it, you now simply pass the reference to the component and you'll have automatically the right type and the mapping name - which leads to the biggest benefit: all methods where components are passed (inside the core) now use this type as return type (before: `GondelComponent`). Other DOM utility methods now support (optional) generic arguments to automatically cast components (before: explicit casting with the `as` operator).

### CODE REVIEW

File: examples/react-lazy/src/components/DemoApp.tsx [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import React, { Component } from 'react';
+import * as React from 'react';
```
Comment (jantimon) [inline examples/react-lazy/src/components/DemoApp.tsx]: Could you please explain why this change is necessary?

File: examples/react-lazy/src/components/DemoApp.tsx [hunk 2]
```diff
@@ -8,9 +8,16 @@ type Props = {
 }
 
 export class App extends React.Component<Props, {}> {
+  onClickButtonNative() {
```
Comment (jantimon) [inline examples/react-lazy/src/components/DemoApp.tsx]: Is this function ever used?

File: examples/react/package.json [hunk 3]
```diff
@@ -11,20 +11,12 @@
   "license": "MIT",
   "dependencies": {
     "@gondel/core": "*",
-    "@gondel/plugin-hot": "*",
     "@gondel/plugin-events": "*",
-    "@types/react": "16.0.40",
```
Comment (jantimon) [inline examples/react/package.json]: Really nice 👍thanks for clean up

File: examples/react/.babelrc [hunk 4]
```diff
@@ -1,17 +0,0 @@
-{
-  "presets": [
```
Comment (jantimon) [inline examples/react/.babelrc]: The idea was to show in this example that Gondel can be used without typescript as well

File: examples/react/src/app.jsx [hunk 5]
```diff
@@ -1,21 +0,0 @@
-import * as React from 'react';
```
Comment (jantimon) [inline examples/react/src/app.jsx]: The idea of app.js was to show an app which is entirely independent from gondel (because it uses just the button wrapper)

Could we please keep it simple here?

File: package.json [hunk 6]
```diff
@@ -12,7 +12,7 @@
     "commitmsg": "conventional-changelog-lint -e",
     "install": "lerna bootstrap --hoist",
     "precommit": "lint-staged",
-    "prettier": "prettier --write **/src/*.ts",
+    "prettier": "prettier --write **/src/*.{ts,tsx}",
```
Comment (jantimon) [inline package.json]: Really nice 👍

File: packages/core/dist/GondelComponent.d.ts [hunk 7]
```diff
@@ -10,6 +17,10 @@ export interface GondelComponent {
     sync?(): void;
 }
 export declare class GondelBaseComponent implements GondelComponent {
+    /**
```
Comment (jantimon) [inline packages/core/dist/GondelComponent.d.ts]: With this change Gondel will only be usable together with a bundler as it enforces to extend from BaseComponent.

I am not 100% sure if this is an issue or not - what do you think?

Also I am not sure if a static value could cause problems with our hot module replacement.  
Do you see an issues there?

File: packages/core/dist/GondelComponent.d.ts [hunk 8]
```diff
@@ -10,6 +17,10 @@ export interface GondelComponent {
...
+     * The components initial identification mappings
```
Comment (jantimon) [inline packages/core/dist/GondelComponent.d.ts]: `The components initial identifcation mappings` seems to be a little bit hard to understand..

What about "An internal mapping to lookup the registered name for each namespace"?

File: packages/core/dist/GondelComponent.d.ts [hunk 9]
```diff
@@ -1,4 +1,11 @@
-export declare type IGondelComponent = new (context: HTMLElement, componentName: string) => GondelComponent;
+export interface IGondelComponent<T = GondelComponent> {
```
Comment (jantimon) [inline packages/core/dist/GondelComponent.d.ts]: If we add the static variable to `GondelBaseComponent` we could switch to `T extends GondelBaseComponent = GondelBaseComponent` so we would no longer need IGondelComponentWithIdentification.

File: packages/core/src/GondelComponentRegistry.ts [hunk 10]
```diff
@@ -38,28 +45,29 @@ export const componentRegistries: {
...
-): void;
-export function registerComponent() {
-  const args = arguments;
-  // The componentName is always the first argument
-  const componentName = args[0] as string;
-  // Use namespace from the second argument or fallback to the default "g" if it is missing
-  const namespace = typeof args[1] === "string" ? args[1] : "g";
-  // The last argument is always the component class
-  let component = args[args.length - 1] as IGondelComponent;
+  component: IGondelComponent,
```
Comment (jantimon) [inline packages/core/src/GondelComponentRegistry.ts]: switching the order would be a breaking change from:

```
registerCompontent('Button', 'g', class Button { ... });
// to
registerCompontent('Button', class Button { ... }, 'g');
```

I believe `Button` and `g` should be close together.  
This change could separate them by hundreds lines of code.

File: packages/core/src/GondelComponentStarter.test.ts [hunk 11]
```diff
@@ -28,8 +28,9 @@ function createMockElement(namespace: string) {
 describe("GondelComponentStarter", () => {
   describe("#component - e2e", () => {
     it("should call the constructor during start", () => {
-      @Component("Button")
```
Comment (jantimon) [inline packages/core/src/GondelComponentStarter.test.ts]: Is this still up-to-date?

File: packages/core/src/GondelDecorators.test.ts [hunk 12]
```diff
@@ -28,24 +24,29 @@ function createMockElement(namespace: string) {
 describe("GondelDecorators", () => {
   describe("#component - e2e", () => {
     it("should add the component for the default 'g' namespace", () => {
-      @Component("Button")
-      class Button extends GondelBaseComponent {}
+      @Component()
+      class Button extends GondelBaseComponent {
+        static componentName = "Button";
```
Comment (jantimon) [inline packages/core/src/GondelDecorators.test.ts]: Is this still up-to-date?

File: packages/core/src/GondelDomUtils.ts [hunk 13]
```diff
@@ -10,6 +15,24 @@ function isElement(domNode: ArrayLikeHtmlElement): domNode is Element {
...
+ * for internal fixtures.
+ *
+ * @param {string=g} namespace
+ * @param {string?} addition
+ * @see https://github.com/ReactiveX/rxjs/blob/master/src/internal/symbol/rxSubscriber.ts
+ */
+export function getGondelAttribute(namespace: string = "g", addition?: string): string {
+  const id = `__gondel_${addition ? addition + "_" : ""}${namespace}__`;
+
+  if (Symbol && typeof Symbol.for === "function") {
```
Comment (jantimon) [inline packages/core/src/GondelDomUtils.ts]: Can we remove the Symbol here?
It doesn't prevent access and will reduce development experience.

File: packages/core/src/GondelDomUtils.ts [hunk 14]
```diff
@@ -25,6 +48,26 @@ export function getFirstDomNode(domNode: ArrayLikeHtmlElement): HTMLElement {
   return domNode[0] as HTMLElement;
 }
 
+/**
+ *
```
Comment (jantimon) [inline packages/core/src/GondelDomUtils.ts]: Please add a description and remove the @param tags

File: packages/core/src/GondelDomUtils.ts [hunk 15]
```diff
@@ -50,69 +93,104 @@ export function stopComponents(domContext?: ArrayLikeHtmlElement, namespace: str
...
 }
 
+export function isComponentMounted(
+  domNode: ArrayLikeHtmlElement,
+  namespace: string = "g"
+): boolean {
+  const firstNode = getFirstDomNode(domNode);
+  const gondelComponent = (firstNode as any)[getGondelAttribute(namespace)];
+
+  return gondelComponent && gondelComponent._ctx;
```
Comment (jantimon) [inline packages/core/src/GondelDomUtils.ts]: this is not of type `boolean` but of type `dom element` so `isComponentMounted(x) === true` would always be false

File: packages/core/src/GondelDomUtils.ts [hunk 16]
```diff
@@ -50,69 +93,104 @@ export function stopComponents(domContext?: ArrayLikeHtmlElement, namespace: str
...
-  return;
+
+  throw new Error(`Component not found in DOM (https://git.io/f4D44).`);
 }
 
 /**
  * Returns the gondel instance for the given HtmlELement once it is booted
  */
-export function getComponentByDomNodeAsync(
+export function getComponentByDomNodeAsync<T extends GondelComponent>(
```
Comment (jantimon) [inline packages/core/src/GondelDomUtils.ts]: Can we rename T to TGondelComponent for better readability?

File: packages/core/src/GondelDomUtils.ts [hunk 17]
```diff
@@ -50,69 +93,104 @@ export function stopComponents(domContext?: ArrayLikeHtmlElement, namespace: str
...
       components.push(gondelComponentInstance);
     }
   }
-  return components;
+
+  if (typeof component === "string") {
+    return components as Array<GondelComponent>;
+  }
+
+  return components as Array<T & GondelComponent>;
```
Comment (jantimon) [inline packages/core/src/GondelDomUtils.ts]: Do we really need ` & GondelComponent` here?

File: packages/plugins/react/src/index.ts [hunk 18]
```diff
@@ -28,6 +28,7 @@ export class GondelReactComponent<S> extends GondelBaseComponent
   _setInternalState: (config: S) => void | undefined;
 
   state: S;
+  __identification = {};
```
Comment (jantimon) [inline packages/plugins/react/src/index.ts]: Do we really need this?
We extend GondelBaseComponent

File: packages/core/src/GondelDomUtils.ts (hunk 17 above)
Comment (janbiasi) [reply packages/core/src/GondelDomUtils.ts]: Yes, I checked how the behaviour would be if we remove GondelComponent, and it looks like that TypeScript won't resolve the BaseComponent type so you wouldn't have access to `_ctx` etc.

**Comment by janwidmer** [comment]
deleted due to inactivity
